DataSetIllegalArgumentException: Base parameter name: "Flow" is not valid on location level getAll

Retrieving location levels for a given office results in the following server-side exception

```
08-Jun-2023 22:19:12.089 WARNING [https-jsse-nio-8443-exec-8] cwms.radar.ApiServlet.lambda$init$15 error on request[3447121928785741572]: /swt-data/levels
	hec.data.DataSetIllegalArgumentException: mil.army.usace.hec.metadata.DataSetIllegalArgumentException: Base parameter name: "Flow" is not valid.
		at hec.data.Parameter.setBaseParameter(Parameter.java:402)
		at mil.army.usace.hec.metadata.Parameter.setParameter(Parameter.java:375)
		at hec.data.Parameter.setParameter(Parameter.java:385)
		at hec.data.Parameter.<init>(Parameter.java:345)
		at hec.data.Parameter.<init>(Parameter.java:333)
		at hec.data.level.JDomLocationLevelRef.<init>(JDomLocationLevelRef.java:258)
		at hec.data.level.JDomLocationLevelImpl.<init>(JDomLocationLevelImpl.java:103)
		at cwms.radar.data.dao.LocationLevelsDaoImpl.toLocationLevel(LocationLevelsDaoImpl.java:394)
		at cwms.radar.data.dao.LocationLevelsDaoImpl.lambda$getLocationLevels$3(LocationLevelsDaoImpl.java:347)
		at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
		at java.util.Iterator.forEachRemaining(Iterator.java:116)
		at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
		at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
		at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
		at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
		at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
		at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
		at cwms.radar.data.dao.LocationLevelsDaoImpl.getLocationLevels(LocationLevelsDaoImpl.java:348)
		at cwms.radar.api.LevelsController.getAll(LevelsController.java:291)
		at io.javalin.apibuilder.CrudFunction$1.invoke$lambda-0(CrudHandler.kt:30)
		at io.javalin.apibuilder.CrudFunctionHandler.handle(CrudHandler.kt)
		at cwms.radar.security.GuestAccessManager.manage(GuestAccessManager.java:21)
		at cwms.radar.security.MultipleAccessManager.manage(MultipleAccessManager.java:41)
		at io.javalin.http.JavalinServlet.addHandler$lambda-0(JavalinServlet.kt:96)
		at io.javalin.http.JavalinServlet$lifecycle$2$1$1.invoke(JavalinServlet.kt:43)
		at io.javalin.http.JavalinServlet$lifecycle$2$1$1.invoke(JavalinServlet.kt:43)
		at io.javalin.http.JavalinServletHandler.executeNextTask(JavalinServletHandler.kt:99)
		at io.javalin.http.JavalinServletHandler.queueNextTaskOrFinish$lambda-1(JavalinServletHandler.kt:85)
		at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:995)
		at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2137)
		at io.javalin.http.JavalinServletHandler.queueNextTaskOrFinish$javalin(JavalinServletHandler.kt:85)
		at io.javalin.http.JavalinServletHandler.executeNextTask$lambda-11$lambda-10(JavalinServletHandler.kt:119)
		at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:616)
		at java.util.concurrent.CompletableFuture.uniApplyStage(CompletableFuture.java:628)
		at java.util.concurrent.CompletableFuture.thenApply(CompletableFuture.java:1996)
		at io.javalin.http.JavalinServletHandler.executeNextTask(JavalinServletHandler.kt:119)
		at io.javalin.http.JavalinServletHandler.queueNextTaskOrFinish$lambda-1(JavalinServletHandler.kt:85)
		at java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:995)
		at java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2137)
		at io.javalin.http.JavalinServletHandler.queueNextTaskOrFinish$javalin(JavalinServletHandler.kt:85)
		at io.javalin.http.JavalinServlet.service(JavalinServlet.kt:89)
		at javax.servlet.http.HttpServlet.service(HttpServlet.java:779)
		at cwms.radar.ApiServlet.service(ApiServlet.java:460)
		at javax.servlet.http.HttpServlet.service(HttpServlet.java:779)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
		at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
		at org.apache.catalina.filters.CorsFilter.handleNonCORS(CorsFilter.java:357)
		at org.apache.catalina.filters.CorsFilter.doFilter(CorsFilter.java:176)
		at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
		at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
		at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197)
		at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
		at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:660)
		at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)
		at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
		at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:687)
		at org.apache.catalina.authenticator.SingleSignOn.invoke(SingleSignOn.java:261)
		at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
		at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360)
		at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399)
		at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
		at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:893)
		at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1789)
		at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
		at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
		at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
		at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
		at java.lang.Thread.run(Thread.java:750)
	Caused by: mil.army.usace.hec.metadata.DataSetIllegalArgumentException: Base parameter name: "Flow" is not valid.
		at mil.army.usace.hec.metadata.Parameter.setBaseParameter(Parameter.java:530)
		at hec.data.Parameter.setBaseParameter(Parameter.java:398)
		... 70 more
```

--- a/cwms-data-api/build.gradle
+++ b/cwms-data-api/build.gradle
@@ -39,7 +39,7 @@ dependencies {
 
     implementation 'org.slf4j:slf4j-jdk14:1.7.36'
 
-    implementation( "mil.army.usace.hec:hec-monolith:2.0.2" ) {
+    implementation( "mil.army.usace.hec:hec-monolith:3.3.20" ) {
         //exclude group: "org.python", module: "jython-standalone"
         //exclude group: "mil.army.usace.hec.swingx"
         exclude group: "*"
@@ -50,8 +50,8 @@ dependencies {
     implementation 'com.google.errorprone:error_prone_annotations:2.15.0'
     runtime 'com.google.flogger:flogger-system-backend:0.7.4'
 
-    implementation "mil.army.usace.hec:hec-nucleus-data:1.0.33"
-    implementation "mil.army.usace.hec:hec-nucleus-metadata:1.0.33"
+    implementation "mil.army.usace.hec:hec-nucleus-data:2.0.1"
+    implementation "mil.army.usace.hec:hec-nucleus-metadata:2.0.1"
     implementation( 'mil.army.usace.hec:hec-cwms-ratings-core:2.0.2' ) {
         
         //exclude group: "mil.army.usace.hec.swingx"

--- a/cwms-data-api/src/main/java/cwms/cda/data/dao/RatingMetadataDao.java
+++ b/cwms-data-api/src/main/java/cwms/cda/data/dao/RatingMetadataDao.java
@@ -47,6 +47,8 @@ import java.sql.Timestamp;
 import java.time.ZonedDateTime;
 import java.util.*;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 import java.util.stream.Collectors;
@@ -62,6 +64,11 @@ public class RatingMetadataDao extends JooqDao<RatingSpec> {
             + "<ratings xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" "
             + "xsi:noNamespaceSchemaLocation=\"https://www.hec.usace.army"
             + ".mil/xmlSchema/cwms/Ratings.xsd\"/>";
+    private final Executor executor = Executors.newCachedThreadPool(r -> {
+        Thread thread = new Thread(r, getClass().getSimpleName());
+        thread.setDaemon(true);
+        return thread;
+    });
 
     private final MetricRegistry metrics;
 
@@ -173,7 +180,7 @@ public class RatingMetadataDao extends JooqDao<RatingSpec> {
             if (useParallel) {
                 mapStream = ratingIds.stream()
                         .map(ratingId -> CompletableFuture.supplyAsync(() ->
-                                retrieveRatings(office, ratingId, start, end)))
+                                retrieveRatings(office, ratingId, start, end), executor))
                         .collect(Collectors.toList())
                         .stream()
                         .map(CompletableFuture::join);

--- a/cwms-data-api/src/test/java/cwms/cda/data/dao/RatingTemplateDaoTest.java
+++ b/cwms-data-api/src/test/java/cwms/cda/data/dao/RatingTemplateDaoTest.java
@@ -1,24 +1,46 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2023 Hydrologic Engineering Center
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
 package cwms.cda.data.dao;
 
-import static cwms.cda.data.dao.DaoTest.getConnection;
-import static cwms.cda.data.dao.DaoTest.getDslContext;
-import static org.junit.jupiter.api.Assertions.assertFalse;
-import static org.junit.jupiter.api.Assertions.assertNotNull;
-import static org.junit.jupiter.api.Assertions.assertTrue;
-
+import cwms.cda.data.dto.rating.RatingTemplate;
 import hec.data.RatingException;
 import hec.data.cwmsRating.RatingSet;
-import hec.data.cwmsRating.io.RatingJdbcCompatUtil;
-import java.io.IOException;
-import java.sql.Connection;
-import java.sql.SQLException;
-import java.util.Set;
+import mil.army.usace.hec.cwms.rating.io.jdbc.RatingJdbcFactory;
 import mil.army.usace.hec.cwms.rating.io.xml.RatingXmlFactory;
 import org.jooq.DSLContext;
 import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
 
-import cwms.cda.data.dto.rating.RatingTemplate;
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Set;
+
+import static cwms.cda.data.dao.DaoTest.getConnection;
+import static cwms.cda.data.dao.DaoTest.getDslContext;
+import static org.junit.jupiter.api.Assertions.*;
 
 
 @Disabled
@@ -93,7 +115,7 @@ class RatingTemplateDaoTest
         RatingSet ratingSet = RatingXmlFactory.ratingSet(xmlRating);
         assertNotNull(ratingSet);
 
-        RatingJdbcCompatUtil.getInstance().storeToDatabase(ratingSet,c, true, true);
+        RatingJdbcFactory.store(ratingSet,c, true, true);
     }
 
 

--- a/cwms-data-api/src/test/java/cwms/cda/data/dao/RatingTemplateDaoTestIT.java
+++ b/cwms-data-api/src/test/java/cwms/cda/data/dao/RatingTemplateDaoTestIT.java
@@ -1,35 +1,52 @@
+/*
+ * MIT License
+ *
+ * Copyright (c) 2023 Hydrologic Engineering Center
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
 package cwms.cda.data.dao;
 
-import static cwms.cda.data.dao.DaoTest.getDslContext;
-import static org.junit.jupiter.api.Assertions.assertFalse;
-import static org.junit.jupiter.api.Assertions.assertNotNull;
-import static org.junit.jupiter.api.Assertions.assertTrue;
-
+import cwms.cda.api.DataApiTestIT;
+import cwms.cda.data.dto.rating.RatingTemplate;
 import fixtures.CwmsDataApiSetupCallback;
-import fixtures.TestAccounts;
 import hec.data.RatingException;
 import hec.data.cwmsRating.RatingSet;
-import hec.data.cwmsRating.io.RatingJdbcCompatUtil;
-import java.io.IOException;
-import java.sql.Connection;
-import java.sql.SQLException;
-import java.util.Set;
-import java.util.function.Consumer;
+import mil.army.usace.hec.cwms.rating.io.jdbc.RatingJdbcFactory;
 import mil.army.usace.hec.cwms.rating.io.xml.RatingXmlFactory;
 import mil.army.usace.hec.test.database.CwmsDatabaseContainer;
 import org.jooq.DSLContext;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.BeforeAll;
-import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Tag;
 import org.junit.jupiter.api.Test;
 
-import cwms.cda.api.DataApiTestIT;
-import cwms.cda.data.dto.rating.RatingTemplate;
+import java.io.IOException;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Set;
+
+import static cwms.cda.data.dao.DaoTest.getDslContext;
+import static org.junit.jupiter.api.Assertions.*;
 
 @Tag("integration")
-@Disabled // tested through RatingControllerIT tests. Currently something about the backwards compatibility module
-// not being found. Suspect class loader issue
 class RatingTemplateDaoTestIT extends DataApiTestIT {
 
     @BeforeAll
@@ -42,8 +59,8 @@ class RatingTemplateDaoTestIT extends DataApiTestIT {
     @AfterAll
     public static void remove_swt_permissiosn() throws Exception {
         CwmsDatabaseContainer<?> databaseLink = CwmsDataApiSetupCallback.getDatabaseLink();
-        removeUserFromGroup(databaseLink.getUsername(), "CWMS Users", "SWT");
-        removeUserFromGroup(databaseLink.getUsername(), "TS ID Creator", "SWT");
+//        removeUserFromGroup(databaseLink.getUsername(), "CWMS Users", "SWT");
+//        removeUserFromGroup(databaseLink.getUsername(), "TS ID Creator", "SWT");
     }
 
     @Test
@@ -122,7 +139,7 @@ class RatingTemplateDaoTestIT extends DataApiTestIT {
         RatingSet ratingSet = RatingXmlFactory.ratingSet(xmlRating);
         assertNotNull(ratingSet);
 
-        RatingJdbcCompatUtil.getInstance().storeToDatabase(ratingSet,c, true, true);
+        RatingJdbcFactory.store(ratingSet,c, true, true);
     }
 
 

